packages/graph-retriever/src/graph_retriever/utils/top_k.py

```diff
@@ -0,0 +1,74 @@
+import heapq
+from typing import Iterable, cast
+
+from graph_retriever.content import Content
+from graph_retriever.utils.math import cosine_similarity_top_k
+
+
+def top_k(batches: Iterable[list[Content]],
+          *,
+          embedding: list[float],
+          k: int,
+          are_batches_sorted: bool = False) -> list[Content]:
+    """
+    Select the top-k contents from the given batches.
+
+    If all contents have scores, they will be used for the comparison rather
+    than being computed.
+
+    Parameters
+    ----------
+    batches : Iterable[list[Content]]
+        The batches of content to select the top-K from.
+    embedding: list[float]
+        The embedding we're looking for.
+    k : int
+        The number of items to select.
+    are_batches_sorted : bool, default False
+        Whether the content of each batch are sorted. If true, and all content
+        has scores, a more efficient top-K selection will be used which doesn't
+        need to consider all of each batch.
+    """
+
+    # TODO: Consider passing threshold here to limit results.
```

Comment:
One optimization that likely makes sense to expose and push down is the score threshold.

If we had a score threshold (say, `score > 0.5`), then we can add that as a filter to every query, meaning stores likely return fewer results (don't need to consider things with lower scores), which could significantly speed up some things. (although, we'd want to consider the case of MMR where a node with a similarity score < 0.5 could have or lead us to a node with an `mmr_score > 0.5`.